### PR TITLE
RandomGeoSampler: several bug fixes

### LINKED ISSUE #408: RandomGeoSampler bias
I've discovered two instances of bias in our current `RandomGeoSampler` and `RandomBatchGeoSampler` implementations.

### Area bias

In our current implementations, we first select a tile uniformly at random, then choose a random patch from that tile. This means that small tiles are just as likely to be sampled as large tiles. For most datasets, this is not an issue, as all tiles are approximately the same size. However, for `IntersectionDatasets`, the area of each intersection can vary widely. Tiles that are barely large enough to return a single sample from will be sampled no less often than massive tiles. This is particularly problematic for `RandomBatchGeoSampler` which would sample an entire batch of images from that very small tile.

This is an issue inherent in our current implementation, but is relatively easy to fix. The solution would be to use a weighted random sampler where weights are derived from the area of each image. Therefore, large tiles will be more likely to be sampled from than small tiles.

### Latitude bias

In certain 2D projections like Mercator, polar regions have the same area as equatorial regions. However, this is not the case on the actual 3D Earth. If we sample at random, we end up oversampling the poles relative to the equator. This has serious consequences for model training including models that are biased towards polar regions or that overestimate/underestimate certain climate patterns.

This is an issue inherent to 2D projections of the Earth in general, not necessarily an issue in TorchGeo. We could still try to do something about this however. One solution would be to force the R-tree index to be in an equal-area projection like Albers. Another solution would be to force the index to be in Mercator and to use a weighted random sampler where the weight comes from (the square root of?) the latitude.

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR includes the following fixes:

- [x] Sample from tiles with larger areas more often (fixes #408)
- [x] Allow `size=0` queries to handle point data
- [x] Allow `size=(height, width)` queries for entire tile at a time
- [x] Use PyTorch's built-in random sampling for easier seed setting

The first requires #375 so has to be in 0.3.0. The rest could theoretically be backported to 0.2.2 but I'm not really planning on a 0.2.2 release at the moment.

### CODE REVIEW

File: torchgeo/samplers/single.py [hunk 1]
```diff
@@ -119,8 +126,9 @@ def __iter__(self) -> Iterator[BoundingBox]:
             (minx, maxx, miny, maxy, mint, maxt) coordinates to index a dataset
         """
         for _ in range(len(self)):
-            # Choose a random tile
-            hit = random.choice(self.hits)
+            # Choose a random tile, weighted by area
+            idx = torch.multinomial(self.areas, 1)
```
Comment (calebrob6) [inline torchgeo/samplers/single.py:130]: this defaults to sampling _without_ replacement

Comment (adamjstewart) [reply torchgeo/samplers/single.py:130]: We only sample a single idx at a time so replacement wouldn't affect anything.

**Review by calebrob6 — APPROVED** [review]
